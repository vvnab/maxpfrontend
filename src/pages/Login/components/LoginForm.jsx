import React, { Component } from 'react';
import PropTypes from 'prop-types';
import validator from 'validator';

import errors from '../../../constants/errors';
import Message from '../../../components/Message';

import styles from './index.module.css';

export class LoginForm extends Component {
  state = {
    email: 'max@test.com',
    password: '12345',
  // email: '',
  // password: ''
  }
  componentWillMount = () => {
    if (this.props.location.pathname === '/logout') {
      this.props.userLogout();
      this.props.history.replace('/login');
    }
  }
  componentWillReceiveProps = (nextProps) => {
    if (nextProps.user.error) {
      this.setState({
        password: ''
      });
    }
    if (!nextProps.user.error && nextProps.user.payload) {
      this.props.history.replace('/profile');
    }
  }
  handleChange = e => {
    const {name, value} = e.target;
    this.setState({
      [name]: value
    });
  }
  handleSubmit = e => {
    e.preventDefault();
    if (!validator.isEmail(this.state.email)) {
      // check email
      this.setState({
        fieldError: true,
        fieldErrorMessage: errors.wrong_email.message
      });
    } else if (this.state.password.length < 5) {
      // check password
      this.setState({
        fieldError: true,
        fieldErrorMessage: errors.wrong_password.message
      });
    } else {
      this.setState({
        fieldError: false,
        fieldErrorMessage: null
      });
      this.props.userLogin(this.state);
    }
  }
  render = () => {
    const {loading, payload, error} = this.props.user;
    const {email, password, fieldError, fieldErrorMessage} = this.state;
    const someError = error || fieldError;
    const errorMessage = fieldErrorMessage || (error && payload);
    return (
      <React.Fragment>
        <section className={ styles.main }>
          <form onSubmit={ this.handleSubmit }>
            <fieldset>
              <legend>Форма входа</legend>
              <p>
                <label>
                  <span className={ styles.label }>Имя:</span>
                  <input type='text' name='email' onChange={ this.handleChange } className={ styles.textField } value={ email } autoComplete='off' />
                </label>
              </p>
              <p>
                <label>
                  <span className={ styles.label }>Пароль:</span>
                  <input type='password' name='password' onChange={ this.handleChange } className={ styles.textField } value={ password } autoComplete='new-password' />
                </label>
              </p>
              <input type="submit" disabled={ loading || !email || !password } value={ loading ? 'Подождите' : 'Вход' } className={ styles.submitButton } />
            </fieldset>
          </form>
        </section>
        { someError && <Message type='error' text={ errorMessage } /> }
        </React.Fragment>
      );
  }
}

LoginForm.propsTypes = {
  user: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired
}

export default LoginForm;