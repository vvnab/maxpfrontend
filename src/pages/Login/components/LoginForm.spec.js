import React from 'React'
import { shallow } from 'enzyme'
import { LoginForm } from './LoginForm'

describe('LoginForm', () => {
  const props = {
    user: {},
    location: {
      pathname: '/logout'
    },
    history: '',
    userLogout: () => {
    },
    userLogin: () => {
    }
  }
  const initialState = {
    email: 'max@test.com',
    password: '12345'
  }

  describe('Login init state', () => {
    const mockUserLogin = jest.fn();
    const nextProps = {
      ...props,
      userLogin: mockUserLogin
    }
    const loginForm = shallow(<LoginForm {...nextProps} />);
    it('init', () => {
      loginForm.setState(initialState);
      expect(loginForm.state()).toEqual(initialState);
    })
  })
  // describe('Login handlers', () => {
  //   const loginForm = shallow(<LoginForm {...nextProps} />);
  //   beforeEach(() => {
  //     loginForm.find('[name="email"]').simulate('change', {

  //     })
  //   })
  //   it('init', () => {
  //     loginForm.setState(initialState);
  //     expect(loginForm.state()).toEqual(initialState);
  //   })
  // })
  describe('Login ok', () => {
    const nextProps = {
      ...props,
    }
    const loginForm = shallow(<LoginForm {...nextProps} />);
    it('renders properly', () => {
      expect(loginForm).toMatchSnapshot();
    })
    it('ok', () => {
      expect(loginForm.find('form')).toHaveLength(1);
      expect(loginForm.find('Message')).toHaveLength(0);
    })
  })
  describe('error', () => {
    const nextProps = {
      ...props,
      user: {
        error: true,
        payload: "Error"
      }
    }
    const loginForm = shallow(<LoginForm {...nextProps} />);
    it('error', () => {
      expect(loginForm.find('Message')).toHaveLength(1);
    })
  })
})

