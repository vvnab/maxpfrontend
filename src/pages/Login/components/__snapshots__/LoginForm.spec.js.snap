// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Login ok renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginForm
    history=""
    location={
        Object {
            "pathname": "/logout",
          }
    }
    user={Object {}}
    userLogin={[Function]}
    userLogout={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <section
          className="main"
>
          <form
                    onSubmit={[Function]}
          >
                    <fieldset>
                              <legend>
                                        Форма входа
                              </legend>
                              <p>
                                        <label>
                                                  <span
                                                            className="label"
                                                  >
                                                            Имя:
                                                  </span>
                                                  <input
                                                            autoComplete="off"
                                                            className="textField"
                                                            name="email"
                                                            onChange={[Function]}
                                                            type="text"
                                                            value="max@test.com"
                                                  />
                                        </label>
                              </p>
                              <p>
                                        <label>
                                                  <span
                                                            className="label"
                                                  >
                                                            Пароль:
                                                  </span>
                                                  <input
                                                            autoComplete="new-password"
                                                            className="textField"
                                                            name="password"
                                                            onChange={[Function]}
                                                            type="password"
                                                            value="12345"
                                                  />
                                        </label>
                              </p>
                              <input
                                        className="submitButton"
                                        disabled={false}
                                        type="submit"
                                        value="Вход"
                              />
                    </fieldset>
          </form>
</section>,
        undefined,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <form
            onSubmit={[Function]}
>
            <fieldset>
                        <legend>
                                    Форма входа
                        </legend>
                        <p>
                                    <label>
                                                <span
                                                            className="label"
                                                >
                                                            Имя:
                                                </span>
                                                <input
                                                            autoComplete="off"
                                                            className="textField"
                                                            name="email"
                                                            onChange={[Function]}
                                                            type="text"
                                                            value="max@test.com"
                                                />
                                    </label>
                        </p>
                        <p>
                                    <label>
                                                <span
                                                            className="label"
                                                >
                                                            Пароль:
                                                </span>
                                                <input
                                                            autoComplete="new-password"
                                                            className="textField"
                                                            name="password"
                                                            onChange={[Function]}
                                                            type="password"
                                                            value="12345"
                                                />
                                    </label>
                        </p>
                        <input
                                    className="submitButton"
                                    disabled={false}
                                    type="submit"
                                    value="Вход"
                        />
            </fieldset>
</form>,
          "className": "main",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <fieldset>
              <legend>
                            Форма входа
              </legend>
              <p>
                            <label>
                                          <span
                                                        className="label"
                                          >
                                                        Имя:
                                          </span>
                                          <input
                                                        autoComplete="off"
                                                        className="textField"
                                                        name="email"
                                                        onChange={[Function]}
                                                        type="text"
                                                        value="max@test.com"
                                          />
                            </label>
              </p>
              <p>
                            <label>
                                          <span
                                                        className="label"
                                          >
                                                        Пароль:
                                          </span>
                                          <input
                                                        autoComplete="new-password"
                                                        className="textField"
                                                        name="password"
                                                        onChange={[Function]}
                                                        type="password"
                                                        value="12345"
                                          />
                            </label>
              </p>
              <input
                            className="submitButton"
                            disabled={false}
                            type="submit"
                            value="Вход"
              />
</fieldset>,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <legend>
                  Форма входа
</legend>,
                <p>
                  <label>
                                    <span
                                                      className="label"
                                    >
                                                      Имя:
                                    </span>
                                    <input
                                                      autoComplete="off"
                                                      className="textField"
                                                      name="email"
                                                      onChange={[Function]}
                                                      type="text"
                                                      value="max@test.com"
                                    />
                  </label>
</p>,
                <p>
                  <label>
                                    <span
                                                      className="label"
                                    >
                                                      Пароль:
                                    </span>
                                    <input
                                                      autoComplete="new-password"
                                                      className="textField"
                                                      name="password"
                                                      onChange={[Function]}
                                                      type="password"
                                                      value="12345"
                                    />
                  </label>
</p>,
                <input
                  className="submitButton"
                  disabled={false}
                  type="submit"
                  value="Вход"
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Форма входа",
                },
                "ref": null,
                "rendered": "Форма входа",
                "type": "legend",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <label>
                    <span
                                        className="label"
                    >
                                        Имя:
                    </span>
                    <input
                                        autoComplete="off"
                                        className="textField"
                                        name="email"
                                        onChange={[Function]}
                                        type="text"
                                        value="max@test.com"
                    />
</label>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <span
                        className="label"
>
                        Имя:
</span>,
                      <input
                        autoComplete="off"
                        className="textField"
                        name="email"
                        onChange={[Function]}
                        type="text"
                        value="max@test.com"
/>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Имя:",
                        "className": "label",
                      },
                      "ref": null,
                      "rendered": "Имя:",
                      "type": "span",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "autoComplete": "off",
                        "className": "textField",
                        "name": "email",
                        "onChange": [Function],
                        "type": "text",
                        "value": "max@test.com",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                  ],
                  "type": "label",
                },
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <label>
                    <span
                                        className="label"
                    >
                                        Пароль:
                    </span>
                    <input
                                        autoComplete="new-password"
                                        className="textField"
                                        name="password"
                                        onChange={[Function]}
                                        type="password"
                                        value="12345"
                    />
</label>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <span
                        className="label"
>
                        Пароль:
</span>,
                      <input
                        autoComplete="new-password"
                        className="textField"
                        name="password"
                        onChange={[Function]}
                        type="password"
                        value="12345"
/>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Пароль:",
                        "className": "label",
                      },
                      "ref": null,
                      "rendered": "Пароль:",
                      "type": "span",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "autoComplete": "new-password",
                        "className": "textField",
                        "name": "password",
                        "onChange": [Function],
                        "type": "password",
                        "value": "12345",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                  ],
                  "type": "label",
                },
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "submitButton",
                  "disabled": false,
                  "type": "submit",
                  "value": "Вход",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "fieldset",
          },
          "type": "form",
        },
        "type": "section",
      },
      undefined,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <section
            className="main"
>
            <form
                        onSubmit={[Function]}
            >
                        <fieldset>
                                    <legend>
                                                Форма входа
                                    </legend>
                                    <p>
                                                <label>
                                                            <span
                                                                        className="label"
                                                            >
                                                                        Имя:
                                                            </span>
                                                            <input
                                                                        autoComplete="off"
                                                                        className="textField"
                                                                        name="email"
                                                                        onChange={[Function]}
                                                                        type="text"
                                                                        value="max@test.com"
                                                            />
                                                </label>
                                    </p>
                                    <p>
                                                <label>
                                                            <span
                                                                        className="label"
                                                            >
                                                                        Пароль:
                                                            </span>
                                                            <input
                                                                        autoComplete="new-password"
                                                                        className="textField"
                                                                        name="password"
                                                                        onChange={[Function]}
                                                                        type="password"
                                                                        value="12345"
                                                            />
                                                </label>
                                    </p>
                                    <input
                                                className="submitButton"
                                                disabled={false}
                                                type="submit"
                                                value="Вход"
                                    />
                        </fieldset>
            </form>
</section>,
          undefined,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <form
              onSubmit={[Function]}
>
              <fieldset>
                            <legend>
                                          Форма входа
                            </legend>
                            <p>
                                          <label>
                                                        <span
                                                                      className="label"
                                                        >
                                                                      Имя:
                                                        </span>
                                                        <input
                                                                      autoComplete="off"
                                                                      className="textField"
                                                                      name="email"
                                                                      onChange={[Function]}
                                                                      type="text"
                                                                      value="max@test.com"
                                                        />
                                          </label>
                            </p>
                            <p>
                                          <label>
                                                        <span
                                                                      className="label"
                                                        >
                                                                      Пароль:
                                                        </span>
                                                        <input
                                                                      autoComplete="new-password"
                                                                      className="textField"
                                                                      name="password"
                                                                      onChange={[Function]}
                                                                      type="password"
                                                                      value="12345"
                                                        />
                                          </label>
                            </p>
                            <input
                                          className="submitButton"
                                          disabled={false}
                                          type="submit"
                                          value="Вход"
                            />
              </fieldset>
</form>,
            "className": "main",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <fieldset>
                <legend>
                                Форма входа
                </legend>
                <p>
                                <label>
                                                <span
                                                                className="label"
                                                >
                                                                Имя:
                                                </span>
                                                <input
                                                                autoComplete="off"
                                                                className="textField"
                                                                name="email"
                                                                onChange={[Function]}
                                                                type="text"
                                                                value="max@test.com"
                                                />
                                </label>
                </p>
                <p>
                                <label>
                                                <span
                                                                className="label"
                                                >
                                                                Пароль:
                                                </span>
                                                <input
                                                                autoComplete="new-password"
                                                                className="textField"
                                                                name="password"
                                                                onChange={[Function]}
                                                                type="password"
                                                                value="12345"
                                                />
                                </label>
                </p>
                <input
                                className="submitButton"
                                disabled={false}
                                type="submit"
                                value="Вход"
                />
</fieldset>,
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <legend>
                    Форма входа
</legend>,
                  <p>
                    <label>
                                        <span
                                                            className="label"
                                        >
                                                            Имя:
                                        </span>
                                        <input
                                                            autoComplete="off"
                                                            className="textField"
                                                            name="email"
                                                            onChange={[Function]}
                                                            type="text"
                                                            value="max@test.com"
                                        />
                    </label>
</p>,
                  <p>
                    <label>
                                        <span
                                                            className="label"
                                        >
                                                            Пароль:
                                        </span>
                                        <input
                                                            autoComplete="new-password"
                                                            className="textField"
                                                            name="password"
                                                            onChange={[Function]}
                                                            type="password"
                                                            value="12345"
                                        />
                    </label>
</p>,
                  <input
                    className="submitButton"
                    disabled={false}
                    type="submit"
                    value="Вход"
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Форма входа",
                  },
                  "ref": null,
                  "rendered": "Форма входа",
                  "type": "legend",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <label>
                      <span
                                            className="label"
                      >
                                            Имя:
                      </span>
                      <input
                                            autoComplete="off"
                                            className="textField"
                                            name="email"
                                            onChange={[Function]}
                                            type="text"
                                            value="max@test.com"
                      />
</label>,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <span
                          className="label"
>
                          Имя:
</span>,
                        <input
                          autoComplete="off"
                          className="textField"
                          name="email"
                          onChange={[Function]}
                          type="text"
                          value="max@test.com"
/>,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Имя:",
                          "className": "label",
                        },
                        "ref": null,
                        "rendered": "Имя:",
                        "type": "span",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "autoComplete": "off",
                          "className": "textField",
                          "name": "email",
                          "onChange": [Function],
                          "type": "text",
                          "value": "max@test.com",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": "input",
                      },
                    ],
                    "type": "label",
                  },
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <label>
                      <span
                                            className="label"
                      >
                                            Пароль:
                      </span>
                      <input
                                            autoComplete="new-password"
                                            className="textField"
                                            name="password"
                                            onChange={[Function]}
                                            type="password"
                                            value="12345"
                      />
</label>,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <span
                          className="label"
>
                          Пароль:
</span>,
                        <input
                          autoComplete="new-password"
                          className="textField"
                          name="password"
                          onChange={[Function]}
                          type="password"
                          value="12345"
/>,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Пароль:",
                          "className": "label",
                        },
                        "ref": null,
                        "rendered": "Пароль:",
                        "type": "span",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "autoComplete": "new-password",
                          "className": "textField",
                          "name": "password",
                          "onChange": [Function],
                          "type": "password",
                          "value": "12345",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": "input",
                      },
                    ],
                    "type": "label",
                  },
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "submitButton",
                    "disabled": false,
                    "type": "submit",
                    "value": "Вход",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "fieldset",
            },
            "type": "form",
          },
          "type": "section",
        },
        undefined,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
